File: Resources/OpenMP_Numerical_Integration-20240130/program_omp_trapezoidal_v1.c
/* OpenMP parallel version of trapezoidal rule */
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#ifdef _OPENMP
#include<omp.h>
#endif

#define PI 3.14159265358

double func(double x)
{
  return (1.0 + sin(x));
}

void trapezoidal_rule(int n, double a, double b, double *result)
{
  double h, x, total; 		/* private  */
  int i;
  int my_rank = omp_get_thread_num(); /* private  */
  int thread_count = omp_get_num_threads();
  int local_n;			
  double local_a, local_b;	/* private scope  */

  h = (b-a)/n;			

  local_n = n/thread_count;	
  local_a = a + my_rank*local_n*h;
  local_b = local_a + local_n*h;
  
  total = (func(local_a) + func(local_b))/2.0;
  for(i = 1; i <= local_n-1; i++) 
    {
      x = local_a + i*h;
      total += func(x);
    }
  total = total * h;

#pragma omp critical
  *result += total; 		/* race condition avoided... */

}


int main(int argc, char* argv[])
{
  double a, b, final_result;
  int n;
  int thread_count = 1;

  if (argc == 2)
    {
      thread_count = strtol(argv[1], NULL, 10);
    }
  else
    {
      printf("\n A command line argument other than name of the executable is required...exiting the program..");
      return 1;
    }

  n = 124;			/* number of trapezoids.. */
  a = 0.0;			/* shared  */
  b = PI;
  final_result = 0.0;		/* shared  */

#pragma omp parallel num_threads(thread_count)
  trapezoidal_rule(n, a, b, &final_result);

  printf("\n The area under the curve (1+sin(x)) between 0 to PI is equal to %lf \n\n", final_result);

  return 0;  
}

File: Resources/OpenMP_Numerical_Integration-20240130/program_omp_trapezoidal_v1.f90
! a little more sophisticated version of hello world
module data
  use OMP_LIB
  implicit none

  double precision, parameter :: PI = 3.14159265358

contains

  double precision function func(x)
    implicit none
    double precision, intent(in) :: x

    func = (1.0 + dsin(x))
    
  end function func

  subroutine trapezoidal_rule(n, a, b, final_result)
    implicit none
    integer, intent(in) :: n
    double precision, intent(in) :: a, b
    double precision, intent(out) :: final_result
    double precision :: h, x, total
    integer :: local_n, i
    double precision :: local_a, local_b
    integer :: my_rank, thread_count

    my_rank = OMP_GET_THREAD_NUM()
    thread_count = OMP_GET_NUM_THREADS()

    h = (b-a)/n
    local_n = n/thread_count;
    local_a = a + my_rank*local_n*h
    local_b = local_a + local_n*h

    total = (func(local_a) + func(local_b))/2.0
    do i = 1, local_n-1
       x = local_a + i*h
       total = total + func(x)
    end do
    total = total*h
    
    !$OMP CRITICAL
    final_result = final_result + total
    !$OMP END CRITICAL

  end subroutine trapezoidal_rule
  
end module data

program main
  use data
  implicit none
  character(100) :: numchar
  character(100) :: name
  double precision :: a, b, final_result
  integer :: n
  integer :: thread_count = 1

  if (COMMAND_ARGUMENT_COUNT() .ne. 1) then
     write(*, *) 'Error, one command line arguments is required, stopping the program'
     STOP
  end if

  call GET_COMMAND_ARGUMENT(0,name)
  write(*, *) name
  
  call GET_COMMAND_ARGUMENT(1, numchar)
  READ(numchar, *) thread_count

  n = 100;
  a = 0.0;
  b = PI
  final_result = 0.0;

  !$OMP PARALLEL NUM_THREADS(thread_count)
  call trapezoidal_rule(n, a, b, final_result)
  !$OMP END PARALLEL

  write(*, *) 'The final result is ', final_result

end program main
